nucypher/blockchain/eth/interfaces.py

```diff
@@ -387,7 +387,9 @@ def _handle_failed_transaction(cls,
 
         try:
             # Assume this error is formatted as an IPC response
-            code, message = exception.args[0].values()
+            response = exception.args[0]
+            code = response['code']
+            message = response['message']
 
         except (ValueError, IndexError, AttributeError):
```

Comment:
This exception handling will need to be updated to consider the case of a `KeyError` (or reworked to exclude) and remove the `ValueError` since this block will no longer do tuple unpacking. If not in this PR than in a follow-up is acceptable, but with a TODO and/or Issue.

Reply:
Motivation for this was exception-during-exception-handler observed when gas estimation failed with geth.

Previous code clearly assumed passed in exception was constructed with a non-empty parameter list of which the first item is a dict or dict-like containing two items the values of which, when iterated over, make sense as 'code' and 'message' in that order.

_handle_failed_transaction() is only called from one site, in build_contract_transaction() in an exception handler surrounding a call to buildTransaction() where it is intended to return an exception that can be raised (i.e. raise _handle_failed_transaction(...)), however _handle_failed_transaction() does not return a value - it always raises an exception itself.

If the intention is for all exceptions during a transaction building to appear as TransactionFailed then the original exception and traceback should be captured and embedded in the TransactionFailed object rather than throwing from the exception handler, along with whatever ad-hoc efforts to extract an error message, etc. from the original exception.

Reply:
For this PR, I've added KeyError and TypeError to the except clause in _handle_failed_transaction().
Arguably this should be a blanket exception clause, as it should just rethrow the original error if it can't decode it.

Reply:
I actually think we're lost meaningful context here by expanding number of handled exceptions.  

- Consider that `Exception.args` is a specification of python exceptions.  If we receive such an object via `AttributeError`, it's an invalid exception.
- A `ValueError` or `TypeError` can no longer caused be explicitly raised in this block and can be removed.  If an implicit exception *does* occur here, it represents an unhanded error can bubble up and raise. 
- `KeyError` indicates that the exception being passed is not formatted as an eth RPC response, and is missing fields.

In order to preserve this context, and eliminate superfluous exception handling I recommend using this approach:

```python
# Extract the exception's message or data.
try:
    response = exception.args[0]
except AttributeError:
    # Python exceptions must have the 'args attribute'.
    raise ValueError(f'{exception} is not a valid Exception instance.')
except IndexError:
    # No exception data or message available.
    raise exception

# Assume this error is formatted as a valid eth RPC response.
try:
    code = response['code']
    message = response['message']
except KeyError:
    # Response data does not contain valid RPC fields.
    raise exception
```

Reply:
Here's how this handler looks in action:  

```
  File "/home/kieran/Git/nucypher/nucypher/cli/commands/stake.py", line 410, in bond_worker
    receipt = STAKEHOLDER.bond_worker(worker_address=worker_address)
  File "/home/kieran/Git/nucypher/nucypher/blockchain/eth/decorators.py", line 116, in wrapped
    return func(actor, *args, **kwargs)
  File "/home/kieran/Git/nucypher/nucypher/blockchain/eth/decorators.py", line 124, in wrapped
    receipt_or_txhash = actor_method(self, *args, **kwargs)
  File "/home/kieran/Git/nucypher/nucypher/blockchain/eth/decorators.py", line 105, in wrapped
    return func(*args, **kwargs)
  File "/home/kieran/Git/nucypher/nucypher/blockchain/eth/actors.py", line 1341, in bond_worker
    receipt = self.staking_agent.bond_worker(staker_address=self.checksum_address,
  File "/home/kieran/Git/nucypher/nucypher/blockchain/eth/decorators.py", line 105, in wrapped
    return func(*args, **kwargs)
  File "/home/kieran/Git/nucypher/nucypher/blockchain/eth/agents.py", line 569, in bond_worker
    receipt: TxReceipt = self.blockchain.send_transaction(contract_function=contract_function,
  File "/home/kieran/Git/nucypher/nucypher/blockchain/eth/decorators.py", line 105, in wrapped
    return func(*args, **kwargs)
  File "/home/kieran/Git/nucypher/nucypher/blockchain/eth/interfaces.py", line 630, in send_transaction
    transaction = self.build_contract_transaction(contract_function=contract_function,
  File "/home/kieran/Git/nucypher/nucypher/blockchain/eth/decorators.py", line 105, in wrapped
    return func(*args, **kwargs)
  File "/home/kieran/Git/nucypher/nucypher/blockchain/eth/interfaces.py", line 503, in build_contract_transaction
    raise self._handle_failed_transaction(exception=error,
  File "/home/kieran/Git/nucypher/nucypher/blockchain/eth/interfaces.py", line 428, in _handle_failed_transaction
    raise transaction_failed from exception
nucypher.blockchain.eth.interfaces.TransactionFailed: BONDWORKER from 0x37f3... 
Sender balance: 0 ETH 
Reason: execution reverted 
Transaction: {'chainId': 4, 'nonce': 0, 'from': '0x37f360567b6C4dF121302DaED8A9B7039FE09Deb'}
```

Reply:
bump @svenski123 (Happy New Year!) - Let's either integrate the changes mentioned above, or replace the PR with another one I can open based on your changes: <https://github.com/nucypher/nucypher/pull/2486#discussion_r548106237>

Reply:
Hi @KPrasch!
Happy belated new year to you as well!
I've rebased and force pushed the branch against main as well as refactored the exception decoding slightly.
In the spirit of paranoia, exception.args[0] can throw a TypeError if .args cannot be indexed, and int(code) where code = exception.args[0]['code'] can throw a ValueError if ...['code'] can't be cast to an int.